### PR TITLE
Workaround for PVR returning wrong uniqueid 

### PR DESCRIPTION
**Pull request by Tolriq**
when asking for Player.GetItem but keep proper value when asking for channel uniqueid

## Description

Fix for https://github.com/xbmc/xbmc/issues/19151

This is mostly a workaround as https://github.com/xbmc/xbmc/pull/18632

## Motivation and Context
Fix for https://github.com/xbmc/xbmc/issues/19151

## How Has This Been Tested?
```
{"id":831,"jsonrpc":"2.0","method":"Player.GetItem","params":{"properties":["album","albumartist","artist","episode","art","file","genre","plot","rating","season","showtitle","studio","tagline","title","track","year","streamdetails","originaltitle","playcount","runtime","duration","cast","writer","director","userrating","firstaired","displayartist","uniqueid"],"playerid":1}}
```

```
{"id":69,"jsonrpc":"2.0","method":"PVR.GetChannels","params":{"channelgroupid":"alltv","properties":["channel","channeltype","hidden","lastplayed","locked","thumbnail","broadcastnow","broadcastnext","channelnumber","subchannelnumber","isrecording","uniqueid"]}}
```

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
@Tolriq doesn't this work?

```diff
diff --git a/xbmc/interfaces/json-rpc/FileItemHandler.cpp b/xbmc/interfaces/json-rpc/FileItemHandler.cpp
index bafbe15aaa..6b3520d06d 100644
--- a/xbmc/interfaces/json-rpc/FileItemHandler.cpp
+++ b/xbmc/interfaces/json-rpc/FileItemHandler.cpp
@@ -77,6 +77,12 @@ bool CFileItemHandler::GetField(const std::string &field, const CVariant &info,
         result[field] = StringUtils::Split(info[field].asString(), EPG_STRING_TOKEN_SEPARATOR);
         return true;
       }
+      else if (field == "uniqueid")
+      {
+        // integer -> Media.UniqueID (string)
+        result[field] = std::to_string(info[field].asInteger());
+        return true;
+      }
     }
   }
 
```

**Comment by ksooo** [comment]
THIS should work:

```diff
diff --git a/xbmc/interfaces/json-rpc/FileItemHandler.cpp b/xbmc/interfaces/json-rpc/FileItemHandler.cpp
index bafbe15aaa..e354514241 100644
--- a/xbmc/interfaces/json-rpc/FileItemHandler.cpp
+++ b/xbmc/interfaces/json-rpc/FileItemHandler.cpp
@@ -77,6 +77,12 @@ bool CFileItemHandler::GetField(const std::string &field, const CVariant &info,
         result[field] = StringUtils::Split(info[field].asString(), EPG_STRING_TOKEN_SEPARATOR);
         return true;
       }
+      else if (field == "uniqueid")
+      {
+        // integer -> Media.UniqueID
+        result[field]["channel"] = std::to_string(info[field].asInteger());
+        return true;
+      }
     }
   }
 
```

**Comment by Tolriq** [comment]
No it's more complex, the same GetField call is used both for Player.GetItem and PVR.GetChannelXXXX
For Player.GetItem it should return an array of value, but those do not exist for PVR and so should return nothing, but since this field exist with another meaning it's currently returned as number and breaks the schema.

But we still need to have normal uniqueid filling when requesting it for channels and it should still be the proper value as of now.

So the patch just intercept the wrong case and fix it, chirurgical workaround.

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp
Comment (phunkyfish) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp:81]: Indentation of brackets is incorrect. Can we also add a comment here detailing why this needs to be here? Otherwise no one will know the purpose.

**Review by phunkyfish — CHANGES_REQUESTED** [review]

**Comment by phunkyfish** [comment]
We would also need to bump json schema version as this would change behaviour.

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp
Comment (ksooo) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp:82]: Please add a remark explaining what exactly is done here and why. Without this, nobody can understand the ratio.

**Review by ksooo — CHANGES_REQUESTED** [review]
This change needs to be explained in the code.

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp
Comment (ksooo) [reply xbmc/interfaces/json-rpc/FileItemHandler.cpp:81]: Ha, @phunkyfish you beat me to this. ;-)

**Comment by Tolriq** [comment]
I'll add a comment tomorrow, but remember this is a quick workaround for v19.
If there's no chance to have this merged then a better solution for v20 is better.

**Comment by ksooo** [comment]
> a quick workaround for v19.

 The more important is the comment. You never know when the right fix will actually find it's way into the code base.

 And is it actually needed for v19? We might break other apps at a very late stage just before releasing v19.

**Comment by phunkyfish** [comment]
> > a quick workaround for v19.
> 
> The more important is the comment. You never know when the right fix will actually find it's way into the code base.
> 
> And is it actually needed for v19? We might break other apps at a very late stage just before releasing v19.

Moreover, is the currently returned value of any real use. I would like to understand what behaviours we might be breaking.

@DaveTBlake for awareness

**Comment by Tolriq** [comment]
> And is it actually needed for v19? We might break other apps at a very late stage just before releasing v19.

PVR is breaking the schema :) Apps are currently broken if they respect it. Most app don't request uniqueID now that I do I'm facing the issue. 
This fix won't break anything at all it fixes a PVR bug.

> Moreover, is the currently returned value of any real use. I would like to understand what behaviours we might be breaking.

PVR currently returns something completely unrelated as uniqueId when something is playing due to a naming clash and how PVR serialize without taking in account impacts on the rest.
This change does not break anything at all it's very very self constrained to fix only the wrongly injected value in one specific case where it breaks the schema and current users.

It's a quick workaround for v19 because of the ```result["channelid"].isNull()``` that is an ugly way to make the difference between requesting json for channels vs currentlyplaying item.
Proper solution would be to expand GetField to have the requested kind as an argument but that's a larger change.

Now Yatse can and will workaround the issue, I lack time to explain how PVR and JSON works more.

**Comment by ksooo** [comment]
> Most app don't request uniqueID

 You never know how many apps are out there relying on the fact uniqueid for channels being an integer. This PR will break those apps, which is not acceptable shortly before v19 release.

> Now Yatse can and will workaround the issue, 

 Great, then we have no urgent need to have this workaround in v19.

> I lack time to explain how PVR and JSON works more.

 Fair enough. With us not understanding the whole picture, another argument for not merging this PR now.

**Comment by Tolriq** [comment]
> You never know how many apps are out there relying on the fact uniqueid for channels being an integer

.... This is the whole point of this PR .... Keeping uniqueid as an integer when requesting it for channels ....

One more reason to never do PR about walled garden parts and leave the bosses fix it.

So I change this to feature request, please stop breaking the schema for v20 :)

**Comment by ksooo** [comment]
> One more reason to never do PR about walled garden parts and leave the bosses fix it.

 Will you ever learn that moaning and offending other people like this will never improve our relationship?

 If you say you have no time to explain things properly, fair enough. But why should we merge something we do not understand?

**Comment by Tolriq** [comment]
This is a JSON side fix that @DaveTBlake would have understood, I opened an issue, was asked to do the research and the PR.
I did and proposed something ultra targeted for v19 inclusion.
The inclusion in v19 is refused, then a proper fix is needed.

Fixing PVR is not something I will risk, so I'm closing this. There's no moaning, just simple facts.

And reason I did so little PR before, as everyone time is the only limited thing that you can't regain, I do not want to loose time to do research and PR that should be discussed before as unwanted and having time lost for nothing.
I had important things to do, I decided to risk to try to have a quick fixe, I failed, I lost a little time no big deal, but I won't lost more.

**Comment by phunkyfish** [comment]
@Tolriq please leave the PR open. 

@DaveTBlake should review it as maintainer of the JSON interface before any decisions are made.

**Comment by DaveTBlake** [comment]
> Great, then we have no urgent need to have this workaround in v19.

@ksooo I do not think that because @Tolriq has a workaround for his app it means we should leave JSON with this issue in v19 when a fix is possible that will not break anything. I would have merged this PR  for v19 once comment and API patch bump added. 

I understand why @Tolriq closed it since from his perspective it was rejected for v19 by the team, and so pointless to leave it open or work on any further. I don't expect him to reopen it, but if he did then I would happily approve it and merge.

I so wish we could all play nicely together. There are so few active devs, and few with deep knowledge of Kodi. Yes people frustrate each other, that includes me, but can we just get over it.

**Comment by Tolriq** [comment]
As title says it was a workaround and not a fix. No need to spend energy and 4 peoples times on this, I'm happy with a Yatse side fix, I stopped counting after 100 ;)

This PR and the related issue have all the data to be able to build a real fix and I'll happily give some knowledge to anyone who want to tackle this.

**Comment by ksooo** [comment]
>  I stopped counting after 100

 I will never get your sense of humor. What do you think this statement will make with the people responsible for the +100 bugs you are suffering from? Please, if you want me to be open and nice, just stop this kind of "humor".

**Comment by DaveTBlake** [comment]
I don't think there is need to take offence here @ksooo. Since Yatse supports old versions of Kodi I'm sure it does have hundreds of workarounds to handle issues that are solved in the newer versions of Kodi.

**Comment by Tolriq** [comment]
Again no humor or anything just a simple fact, and yes quite a few have been fixed, I even have some fixes made by me and pushed by other users..... One major one by you recently about the exact same issue here for other fields.

I did try for years to have discussions with you about PVR and JSON support lacking stuff and having issues.
I was glad to see recently some movement and a very long lasting bug fixed, but how hard to get you accept an issue that I was able to reproduce as the demo PVR addon did not produce the issue. How many don't care about what you say, open an issue with logs?

Now about this, I opened an issue, got an answer no time do it. I also have not, but took time to try due to soon v19 final release, first thing you do, is here's other ways without even trying to understand the work and research already done, maybe that's not the best approach to encourage other people to contribute? And maybe that's a root problem way bigger than just me and my sense of humor.

**Pull request closed without merging**
